BlockChainRunner.java

```diff
@@ -0,0 +1,354 @@
+import static com.karolismed.simple_blockchain.utils.StringHelper.log;
+import static java.util.Objects.isNull;
+
+public class BlockChainRunner {
+
+    private final BlockChainSetupService setupService;
+    private final MerkleTreeConstructor merkleTreeConstructor;
+
+    private List<User> users;
+    private List<Block> blockChain;
+    private List<Transaction> transactionPool;
+    private List<Transaction> pendingTransactions;
+    private MultiValuedMap<PublicKey, UnspentOutput> unspentOutputMap;
+    private Map<String, Integer> txIdBlockIndexMap;
+    private Map<PublicKey, User> pkUserMap;
+
+    private int totalTransactionsPooled;
+
+    public BlockChainRunner() {
+        setupService = new BlockChainSetupService();
+        merkleTreeConstructor = new MerkleTreeConstructor();
+
+        blockChain = new LinkedList<>();
+        users = new ArrayList<>();
+        transactionPool = new LinkedList<>();
+        pendingTransactions = new ArrayList<>();
+        unspentOutputMap = new HashSetValuedHashMap<>();
+        txIdBlockIndexMap = new HashMap<>();
+        pkUserMap = new HashMap<>();
+    }
+
+    public void init() throws Exception {
+        log("Setting up BlockChain...");
+        setupService.setupUsers(users, pkUserMap);
+
+        log("Mining genesis block...");
+        mineGenesisBlock();
+
+        log("Setting up miners...");
+        BlockingQueue<Pair<Block, Integer>> blockQueue = new LinkedBlockingQueue<>();
+        Semaphore threadSyncSemaphore = new Semaphore(11);
+        Miner[] miners = {
+            new Miner("Miner 1", transactionPool, 0, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 2", transactionPool, 1, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 3", transactionPool, 2, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 4", transactionPool, 3, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 5", transactionPool, 4, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 6", transactionPool, 5, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 7", transactionPool, 6, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 8", transactionPool, 7, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 9", transactionPool, 8, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 10", transactionPool, 9, blockQueue, threadSyncSemaphore),
+            new Miner("Miner 11", transactionPool, 10, blockQueue, threadSyncSemaphore)
+        };
+        startMinerThreads(miners);
+
+
+        while (true) {
+            generatePendingTransactions();
+            addPendingTransactionsToPool();
+
+            if (transactionPool.isEmpty()) break;
+
+            threadSyncSemaphore.acquire(miners.length);
+            blockQueue.clear();
+            startMiningNextBlock(blockChain.get(blockChain.size() - 1).getHash(), REGULAR_BLOCK_DIFFICULTY, miners);
```

Comment:
Seems to be a violation of Law of Demeter, since blockChain.get().getHash() is breaking the initial principle.

Would be suitable to either reform the startMiningNextBlock to accept just the blockChain.get() output, or put the output of blockChain.get() to a variable.

Reply:
Addressed